### ISSUE TITLE
Negative indexing for nn.Embedding inputs

### ISSUE THREAD

**Issue by j-min** [pytorch/pytorch#14844]
## 🚀 Feature

Negative indexing for nn.Embedding inputs

## Motivation

I found both
- `nn.Embedding`
- `nn.utils.rnn.pad_sequence`

allow negative indexing for padding.

How about we allow negative indexing for `nn.Embedding` inputs as well?

## Example

```
from torch import nn
from torch.nn.utils.rnn import pad_sequence
import torch

input = [
    torch.LongTensor([1,2,3]),
    torch.LongTensor([1,4]),
    torch.LongTensor([2])]

padded_input = pad_sequence(input, batch_first=True, padding_value=-1)
# tensor([[ 1,  2,  3],
#         [ 1,  4, -1],
#         [ 2, -1, -1]])

embedding = nn.Embedding(3, 10, padding_idx=-1)

embedded = embedding(padded_input) <= error (index out of range...)
```

**Comment by zou3519** [comment]
Maybe replacing index_select with indexing is sufficient

**Comment by Rishav09** [comment]
Hi, I am facing the same Runtime error while running a code ,I am not able to understand the problem and how to solve it.

**Comment by albanD** [comment]
@zou3519 But is that the expected behavior here? I am not sure about this but in the above example, if you wrap negative values, then 4 and -1 become the same thing. But -1 should be padding and 4 is an input and so they should be different.
Isn't the problem just that the `padding_idx` argument is ignored in the forward function [here](https://github.com/pytorch/pytorch/blob/517c7c98610402e2746586c78987c64c28e024aa/aten/src/ATen/native/Embedding.cpp#L17-L32) ?


@Rishav09 The problem is that one of the element you give as input to the embedding is larger than the size of the embedding. You need to check that your inputs are correct: between 0 and input_size-1.

**Comment by albanD** [comment]
Even though I guess that could be confusing, it can be interesting to add.
The goal would just be to add support for negative indices either by replacing the `index_select` in the cpp by something that allows negative indices. Or preprocess the indices before giving them to `index_select` to ensure they are all in the correct range.

**Comment by ksasso1028** [comment]
sup wit this?
